### PR TITLE
feat: add custom Azure host/port to support custom blob endpoint

### PR DESCRIPTION
**Pull request by jeqo**
# About this change - What it does

Adds custom host and port configs to Azure to support custom blob endpoint like Azurite.

Resolves: #163 

# Why this way

Follow similar approach to S3, to set custom host/port.
Moved conn_string construction out of the initialization to be able to test it.

### CODE REVIEW

File: rohmu/object_storage/azure.py
Comment (rominf) [inline rohmu/object_storage/azure.py:139]: Is it debugging leftovers?

File: rohmu/object_storage/azure.py
Comment (rominf) [inline rohmu/object_storage/azure.py:132]: Let's assume `is_secure == False`, but `host` and `port` are `None`. Returning connection string will be `https`, which is unexpected.

File: rohmu/object_storage/azure.py
Comment (rominf) [inline rohmu/object_storage/azure.py:150]: There is duplication of 3/4 lines of `conn_str`. I think this can be rewritten to be readable and without duplication the same time.

File: test/object_storage/test_azure.py
Comment (rominf) [inline test/object_storage/test_azure.py:108]: Since there is a strange behavior on I pointed out above, I suggest to write tests without custom host and port as well.

**Review by rominf — CHANGES_REQUESTED** [review]

File: rohmu/object_storage/azure.py
Comment (jeqo) [reply rohmu/object_storage/azure.py]: ofc 😅 , removed in last commit

File: rohmu/object_storage/azure.py
Comment (jeqo) [reply rohmu/object_storage/azure.py]: Sure. Wasn't considering it for default endpoint, but there may be some scenario where this is needed. Refactored in last commit

**Review by rominf — APPROVED** [review]

**Review by exg — CHANGES_REQUESTED** [review]
Fixup commits should be folded.

**Comment by giacomo-alzetta-aiven** [comment]
Overall LGTM but I agree with Emanuele that commit history could be cleaned up before merging.

**Comment by jeqo** [comment]
@exg @giacomo-alzetta-aiven thanks! Commits are folded now, have another look.

File: rohmu/object_storage/azure.py
Comment (exg) [inline rohmu/object_storage/azure.py:140]: It would be better to use pydantic validators for custom validation, see https://docs.pydantic.dev/1.10/usage/validators/

Comment (jeqo) [reply rohmu/object_storage/azure.py]: Good idea. Adding this on the latest commit, have a look

Comment (exg) [reply rohmu/object_storage/azure.py]: The validation of `azure_cloud` can also be moved in the validator. Also, remember to fold fixup commits before pushing :)

Comment (jeqo) [reply rohmu/object_storage/azure.py]: Sure, can move it to validators as well.

Re: fixups, I prefer to keep them separate for easier review, but can surely rebase/squash once approved if that's ok.

Comment (exg) [reply rohmu/object_storage/azure.py]: That is also fine, and I used to recommend that too before. I just find it unnecessary given that GitHub provides the diff between the previous and current branch tip on force push (via the "Compare" button).

Comment (giacomo-alzetta-aiven) [reply rohmu/object_storage/azure.py]: Also if a reviewer marks the files as "viewed" when reviewing GitHub will also show them which files have changed when the PR is updated, so it's not like reviewers are forced to re-read everything to discover what you updated.

Comment (jeqo) [reply rohmu/object_storage/azure.py]: Thanks, will keep that in mind. 
I've squashed all commits now for this change.

File: rohmu/object_storage/config.py
Comment (exg) [inline rohmu/object_storage/config.py:102]: This test is redundant because the fields have a default value and so they are always present. BTW, another option would be to introduce a `HostInfo` model, similar to `ProxyInfo`, containing the new fields. This way there would be no need for a custom validator, and it would be more consistent. WDYT?

Comment (jeqo) [reply rohmu/object_storage/config.py]: Good catch, fixing that on the latest commit. 

About HostInfo, I like the idea, but we could do that as a separate refactoring PR as there are many places where this type would fit. At the moment this is consistent with the S3 approach to set custom host/port.

Comment (exg) [reply rohmu/object_storage/config.py]: Agreed, I did not realize that this is based on the existing S3 config.

File: rohmu/object_storage/config.py
Comment (exg) [inline rohmu/object_storage/config.py:102]: This rejects (existing) configs where host and port are both missing, which is not correct. I think you want
```
(values["host"] is None) != (values["port"] is None)
```

Comment (giacomo-alzetta-aiven) [reply rohmu/object_storage/config.py:102]: Yeah, `test_azure_config_host_port_set_together` or a similar test should also verify that the valid cases do not raise an error.

Comment (jeqo) [reply rohmu/object_storage/config.py]: Oh, good catch! I added tests to validate the minimal case and fix this.

**Review by exg — APPROVED** [review]

File: rohmu/object_storage/azure.py
Comment (exg) [reply rohmu/object_storage/azure.py:128]: So it is possible to create transfers bypassing the pydantic config models?

Comment (giacomo-alzetta-aiven) [reply rohmu/object_storage/azure.py:128]: I'm just saying that we are not indicating that `AzureTransfer` is private, so users could instantiate it directly instead of going through the factory methods that use `from_model` to create the instance from the pydantic configuration.

I don't this is a big deal, hence why I class this as nitpicking.

**Comment by giacomo-alzetta-aiven** [comment]
Since we all agree let's merge this.

**Pull request merged**
